cub/cub/device/dispatch/dispatch_streaming_reduce.cuh

```diff
@@ -0,0 +1,451 @@
+    PromoteToGlobalOpT promote_op,
+    GlobalReductionOpT reduce_op)
+      : last_partition(is_last_partition)
+      , d_previous_aggregate(d_previous_aggregate)
+      , d_aggregate_out(d_aggregate_out)
+      , d_out(d_out)
+      , promote_op(promote_op)
+      , reduce_op(reduce_op)
+  {}
+
+  template <typename IndexT, typename AccumT>
+  _CCCL_DEVICE _CCCL_FORCEINLINE void operator()(IndexT, AccumT per_partition_aggregate)
+  {
+    // Add this partitions aggregate to the global aggregate
+    if (first_partition)
+    {
+      *d_aggregate_out = promote_op(per_partition_aggregate);
+    }
+    else
+    {
+      *d_aggregate_out = reduce_op(*d_previous_aggregate, promote_op(per_partition_aggregate));
+    }
+
+    // If this is the last partition, we write the global aggregate to the user-provided iterator
+    if (last_partition)
+    {
+      *d_out = *d_aggregate_out;
+    }
+  }
+
+  /**
+   * This is a helper function that's invoked after a partition has been fully processed
+   */
+  template <typename GlobalOffsetT>
+  _CCCL_HOST_DEVICE _CCCL_FORCEINLINE void advance(GlobalOffsetT partition_size, bool next_partition_is_the_last)
+  {
+    promote_op.advance(partition_size);
+    ::cuda::std::swap(d_previous_aggregate, d_aggregate_out);
+    first_partition = false;
+    last_partition  = next_partition_is_the_last;
+  }
+};
+
+/**
+ * Unary promotion operator type that is used to transform a per-partition result to a global result
+ */
+template <typename GlobalOffsetT>
+struct promote_to_global_op
+{
+  // The current partition's offset to be factored into this partition's index
+  GlobalOffsetT current_partition_offset;
+
+  /**
+   * This helper function is invoked after a partition has been fully processed, in preparation for the next partition.
+   */
+  _CCCL_HOST_DEVICE _CCCL_FORCEINLINE void advance(GlobalOffsetT partition_size)
+  {
+    current_partition_offset += partition_size;
+  }
+
+  /**
+   * Unary operator called to transform the per-partition aggregate of a partition to a global aggregate type (i.e., one
+   * that is used to reduce across partitions).
+   */
+  template <typename PerPartitionOffsetT, typename AccumT>
+  _CCCL_HOST_DEVICE _CCCL_FORCEINLINE KeyValuePair<GlobalOffsetT, AccumT>
+  operator()(KeyValuePair<PerPartitionOffsetT, AccumT> partition_aggregate)
+  {
+    return KeyValuePair<GlobalOffsetT, AccumT>{
+      current_partition_offset + static_cast<GlobalOffsetT>(partition_aggregate.key), partition_aggregate.value};
+  }
+};
+
+/**
+ * Offsets a the given input iterator by a fixed offset, such that when an item at index `i` is accessed, the item
+ * `it[*offset_it + i]` is accessed.
+ */
+template <typename Iterator, typename OffsetItT>
+class offset_iterator : public THRUST_NS_QUALIFIER::iterator_adaptor<offset_iterator<Iterator, OffsetItT>, Iterator>
+{
+public:
+  using super_t = THRUST_NS_QUALIFIER::iterator_adaptor<offset_iterator<Iterator, OffsetItT>, Iterator>;
```

Comment:
`offset_iterator` here already referes to the current template instantiation (called the "injected class name"), so no need to supply template arguments.
```suggestion
  using super_t = THRUST_NS_QUALIFIER::iterator_adaptor<offset_iterator, Iterator>;
```